Create article - pre-select language & access level

pe7er opened the pull request:
This PR fixes the **Article Edit form** with **pre-selection** of **Language** and **Access Level**.
Issue: if you want to add a lot of new articles and assign them to other Languages or Access Levels,
you've to select those manually in each new article. 
## Test instructions

Go to Article Manager
In back-end > Content > Articles > [Search Tools]
**Set Filters** for **Select Category**, **Select Language** (to something different than "All") and **Select Access** (to something different than "Public").

![screen shot 2015-05-17 at 09 48 27](http://issues.joomla.org/uploads/1/7c46214732996e045273e8b206fdb5fa.png)

Create a **New Article**
Look on the right:
- Category -> the Category that you've filtered on in the Article Manager
- Access -> **always "Public"**
- Language -> **always "All"**

![screen shot 2015-05-17 at 09 48 27](http://issues.joomla.org/uploads/1/6a03175829c8040b13de4f20fd43d974.png)
### This PR fixes the "pre-selection"

on basis of Access Level & Language filters. 
After the PR the parameters on the right should show:
- Category -> the Category that you've filtered on in the Article Manager
- Access -> the Access Level that you've filtered on in the Article Manager
- Language -> the Language that you've filtered on in the Article Manager

![screen shot 2015-05-17 at 09 48 27](http://issues.joomla.org/uploads/1/f22f66853ba38a75bde1f488163bdbe1.png)

pe7er committed:
Create article - pre-select language & access level

bertmert commented:
@test Success. Hathor, too.

@pe7er Nice feature. I like it.

Bakual commented on administrator/components/com_content/models/article.php:
```diff
@@ -426,6 +426,14 @@ protected function loadFormData()
+				$data->set('language', $app->input->getVar('language', $filterLanguage));
+
+				// Pre-select Access Level if set in Article Manager: Articles > Search Tools > Set Access
+				$filterLanguage = isset($filters['access']) ? $filters['access'] : null;
```
Can you please rename this variable to a more matching name like `$filterAccess`? Thanks.

gunjanpatel commented:
All is working fine. Great addition. Thanks Peter.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6966">issues.joomla.org/joomla-cms/6966</a>.</sub>

pe7er committed:
Renamed one pre-select variable to $filterAccess

gunjanpatel commented:
Also tested successfully after code changes in variable name.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6966">issues.joomla.org/joomla-cms/6966</a>.</sub>

infograf768 commented:
Should not we pre-select language only when JLanguageMultilang::isEnabled()  ?

pe7er commented:
Thank you all for testing! Thank you @Bakual for the code improvement suggestion. Well spotted! I've changed it in the PR.

@wilsonge Why was this PR added to milestone 3.5.0? 
IMHO it's not a new feature: 
Pre-selection was working for Category, but not for Language & Access Level. This PR fixed that issue. 
Can this fix be added to 3.4.2 instead? Thanks!

pe7er commented:
Could you please explain the benefit your suggestion @infograf768 ?

IMHO it's not needed to check for JLanguageMultilang::isEnabled()
* What if admins are creating multilingual articles for a new language, 
before wanting to switch on the Multilingal feature in the Language Plugin?
* Wouldn't it be confusing for admins that the option sometimes works and sometimes not?

infograf768 commented:
>What if admins are creating multilingual articles for a new language, before wanting to switch on the Multilingal feature in the Language Plugin?

This makes sense.

sovainfo commented:
No need to waste resources, use 1 variable ($filterAttribuut) if you have to. You don't have to:
<code>
// Prime some default values.
			if ($this->getState('article.id') == 0)
			{
				$filters = (array) $app->getUserState('com_content.articles.filter');
				$data->set('catid', $app->input->getInt('catid', isset($filters['category_id']) ? $filters['category_id'] : null));
				$data->set('language', $app->input->getVar('language', isset($filters['language']) ? $filters['language'] : null));
				$data->set('access', $app->input->getInt('access', isset($filters['access']) ? $filters['access'] : null));
			}
</code>

pe7er committed:
Rewritten the $data->set lines

pe7er commented:
Thanks @sovainfo !
I've implemented your suggestion. 
Note that you have to put the "isset" part between brackets. I changed your
$data->set('catid', $app->input->getInt('catid', isset($filters['category_id']) ? $filters['category_id'] : null));
to
$data->set('catid', $app->input->getInt('catid', **(**isset($filters['category_id']) ? $filters['category_id'] : null**)**));
to get it working.

pe7er committed:
Added "Status" to pre-selectable fields

pe7er commented:
I've added the **Status field** to the pre-selectable fields. **Please re-test!**

## Testing procedure
Go to Article Manager. In back-end > Content > Articles > [Search Tools]
**Set Filters** for 
Select **Status** (something else than "Published"),
Select **Category**, 
Select **Language** (to something different than "All") and
Select **Access** (to something different than "Public").

![screen shot 2015-05-18 at 03 38 15](http://issues.joomla.org/uploads/1/ac126e7093b25d399937896e0e07f8aa.png)

Create a **New Article** and check if the following parameters are automatically selected:
Status -> the Status that you've filtered on in the Article Manager
Category -> the Category that you've filtered on in the Article Manager
Access -> the Access Level that you've filtered on in the Article Manager
Language -> the Language that you've filtered on in the Article Manager

![screen shot 2015-05-18 at 03 38 15](http://issues.joomla.org/uploads/1/dbcbe73b283c2da829390ae9532b1cfa.png)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6966">issues.joomla.org/joomla-cms/6966</a>.</sub>

infograf768 commented:
Not sure about the status field, Peter. One rarely creates a new item with a different status than published. Implementing this may confuse users.

pe7er commented:
I've customers who want to create multiple new articles, but do not want to publish them right away. For them this feature will be useful. 
And IMHO it will be more consistent to have the same behavior on the "Status" filter as on the Language & Access filters.

brianteeman commented:
Yes PLEASE on the status field

On 18 May 2015 at 10:41, Peter Martin <notifications@github.com> wrote:

> I've customers who want to create multiple new articles, but do not want
> to publish them right away. For them this feature will be useful.
> And IMHO it will be more consistent to have the same behavior on the
> "Status" filter as on the Language & Access filters.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/6966#issuecomment-102996988>.
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

sovainfo commented:
@pe7er Strange, worked for me without them. w8.1 xampp 1.8.1 PHP 5.4.7

To be consistent, any field that makes sense. Don't consider other fields of the filter making sense.

@test Still OK

brianteeman commented:
Two goo tests of the final version of the code - setting RTC - thanks all<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6966">issues.joomla.org/joomla-cms/6966</a>.</sub>

infograf768 commented:
`$data->set('language', $app->input->getVar('language', (isset($filters['language']) ? $filters['language'] : null)));`

should be changed to
`$data->set('language', $app->input->getString('language', (isset($filters['language']) ? $filters['language'] : null)));`

Similar error corrected in: https://github.com/joomla/joomla-cms/pull/7039 
Please take off RTC status.

zero-24 commented:
@pe7er I hope you can have a look into this after JAB or maybe tomorrow on the `Hake it happen` Session :smile:

wilsonge commented:
Merged including the change suggested by @infograf768
